device/transport/mqtt/src/mqtt.ts

```diff
@@ -547,4 +608,222 @@ export class Mqtt extends EventEmitter implements Client.Transport, StableConnec
+            msg.properties.add(k, v);
+            break;
+        }
+      }
+    }
+
+    /*Codes_SRS_NODE_DEVICE_MQTT_RECEIVER_16_004: [If there is a listener for the message event, a message event shall be emitted for each message received.]*/
+    this.emit('message', msg);
+  }
+
+  private _onDeviceMethod(topic: string, payload: any): void {
+    // The topic name looks like this:
+    //  $iothub/methods/POST/{method name}?$rid={request id}&{serialized properties}
+    // We parse out the message.
+
+    /* Codes_SRS_NODE_DEVICE_MQTT_RECEIVER_13_005: [ When a method_<METHOD NAME> event is emitted the parameter shall conform to the shape as defined by the interface specified below:
+
+      interface StringMap {
+          [key: string]: string;
+      }
+
+      interface MethodMessage {
+          methods: { methodName: string; };
+          requestId: string;
+          properties: StringMap;
+          body: Buffer;
+          verb: string;
+      }
+    ]*/
+    const methodMessage = _parseMessage(topic, payload);
+    if (!!methodMessage) {
+      // Codes_SRS_NODE_DEVICE_MQTT_RECEIVER_13_003: [ If there is a listener for the method event, a method_<METHOD NAME> event shall be emitted for each message received. ]
+      // we emit a message for the event 'method_{method name}'
+      this.emit('method_' + methodMessage.methods.methodName, methodMessage);
+    }
+  }
+}
+
+/**
+ * @private
+ */
+interface TopicDescription {
+  name: string;
+  listenersCount: number;
+  subscribeInProgress: boolean;
+  subscribed: boolean;
+  topicMatchRegex: RegExp;
+  handler: Function;
+}
+
+/**
+ * @private
+ */
+class MethodDescription {
+  methodName: string;
+  verb: string;
+}
+
+/**
+ * @private
+ */
+class MethodMessage {
+  methods: MethodDescription;
+  requestId: string;
+  properties: { [key: string]: string };
+  body: Buffer;
+}
+
+function _parseMessage(topic: string, body: any): MethodMessage {
+  let url: URL.Url;
+  let path: string[];
+  let query: any;
+  try {
+    url = URL.parse(topic);
+    path = url.path.split('/');
+    query = querystring.parse(url.query);
+  } catch (err) {
+    return undefined;
+  }
+
+  // if the topic has a querystring then 'path' will include it; so
+  // we strip it out
+  const lastPathComponent = path[path.length - 1];
+  if (lastPathComponent.indexOf('?') !== -1) {
+    path[path.length - 1] = lastPathComponent.substr(
+      0, lastPathComponent.indexOf('?')
+    );
+  }
+
+  if (path.length > 0 && path[0] === '$iothub') {
+    let message = new MethodMessage();
+    if (path.length > 1 && path[1].length > 0) {
+      // create an object for the module; for example, $iothub/twin/...
+      // would result in there being a message.twin object
+      let mod = message[path[1]] = new MethodDescription();
+
+      // parse the request ID if there is one
```

Comment:
>parse the request ID if there is one [](start = 9, length = 36)

heh?  doesn't look like it's parsing to me.

Reply:
haha fair. changed the comment.

---
In reply to: [141921865](https://github.com/Azure/azure-iot-sdk-node/pull/96#discussion_r141921865) [](ancestors = 141921865)